### PR TITLE
[RAC-6211] Code Refine and Unit test for WSMAN inventory job

### PR DESCRIPTION
**Pull request by AlaricChan**
**Background**
Based on the preceding investigation work, start to port redfish get inventory underlying implementation from Racadm with WSMAN API.

**Details**
1. Code refine in WSMAN inventory job
2. Add unit tests for WSMAN inventory job

**Reviewer**
@leoyjchang @nortonluo @lanchongyizu @yaolingling @anhou

### CODE REVIEW

**Comment by yaolingling** [comment]
test this please

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        it('should handle the case of lacking dell configuration.', function() {
+            wsmanBaseJob.prototype.checkOBM.resolves(obm);
+            configuration.get.withArgs('dell').returns(undefined);
+            expect(wsmanInventoryJob._initJob()).to.be.rejectedWith('Dell Inventory web service is not defined in wsmanConfig.json.');
+        });
+
+        it('should handle the case of incorrect obm setting that ip address is undefined.', function() {
+            wsmanBaseJob.prototype.checkOBM.resolves(obm);
+            wsmanBaseJob.prototype.getIpAddress.resolves(undefined);
+            configuration.get.withArgs('dell').returns(dellConfigs);
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: move line 85 to line 84

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+            expect(wsmanInventoryJob._initJob()).to.be.rejectedWith('No target IP address.');
+        });
+
+        it('should handle the case of incorrect node type', function() {
+            wsmanInventoryJob.nodeType = 'unknown';
+            wsmanBaseJob.prototype.checkOBM.resolves(obm);
+            wsmanBaseJob.prototype.getIpAddress.resolves('172.31.128.73');
+            configuration.get.withArgs('dell').returns(dellConfigs);
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: the same problem

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            expect(wsmanInventoryJob._initJob()).to.be.rejectedWith('Inventory collection for node type (unknown) is not implemented.');
+        });
+
+        it('should init job correctly', function() {
+            wsmanInventoryJob.nodeType = 'compute';
+            wsmanBaseJob.prototype.checkOBM.resolves(obm);
+            wsmanBaseJob.prototype.getIpAddress.resolves('172.31.128.73');
+            configuration.get.withArgs('dell').returns(dellConfigs);
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: the same problem

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob._initJob().then(function() {
+                expect(wsmanInventoryJob.retrys.length === 4);
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: retrys? It should be inventories and It's better to add value identify. For example, 
expect(wsmanInventoryJob.inventories[0]).to.euqal('hardware')

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            });
+        });
+    });
+
+    describe('inventoryCallback', function() {
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: It's better to remove test case for inventoryCallback function. It doesn't make sense to add test case for it.

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                elementName: {
+                    value: 'Integrated Dell Remote Access Controller',
+                    otherAttributes: {}
+                }
+            }
+        };
+
+        beforeEach(function() {
+            loggerSpy.reset();
+            WsmanInventoryJobClass.prototype.handleAsyncResponse.restore();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: what's purpose?

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        });
+
+        it('should handle the case that no catalog created.', function() {
+            return wsmanInventoryJob.handleAsyncResponse({}, 'software').then(function() {
+                expect(loggerSpy).to.have.been.calledOnce;
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: It can't identify which case is running.

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            });
+        });
+
+        it('should handle the case that no catalog retrieved with retry.', function() {
+            wsmanInventoryJob.retrys = ['hardware', 'manager'];
+            return wsmanInventoryJob.handleAsyncResponse({}, 'software').then(function() {
+                expect(loggerSpy).to.have.been.calledOnce;
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: same problem

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+        });
+
+        it('should handle the case that no catalog created.', function() {
+            return wsmanInventoryJob.handleAsyncResponse({}, 'software').then(function() {
+                expect(loggerSpy).to.have.been.calledOnce;
+            });
+        });
+
+        it('should handle the case that no catalog retrieved with retry.', function() {
+            wsmanInventoryJob.retrys = ['hardware', 'manager'];
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: retrys? It maybe 'inventories'

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob.handleAsyncResponse({}, 'software').then(function() {
+                expect(loggerSpy).to.have.been.calledOnce;
+            });
+        });
+
+        it('should create new inventory catalog.', function() {
+            wsmanInventoryJob.retrys = ['hardware', 'software', 'manager'];
+            mockWaterline.catalogs.findLatestCatalogOfSource.resolves(catalog);
+            mockWaterline.catalogs.updateByIdentifier.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: remove this line

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob.handleAsyncResponse(data, 'software').then(function() {
+                expect(mockWaterline.catalogs.updateByIdentifier).to.have.been.calledOnce;
+            });
+        });
+
+        it('should update existed inventory catalog.', function() {
+            wsmanInventoryJob.retrys = ['hardware', 'software', 'manager'];
+            mockWaterline.catalogs.findLatestCatalogOfSource.resolves(undefined);
+            mockWaterline.catalogs.create.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: remove this line

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob.handleAsyncResponse(data, 'nics').then(function() {
+                expect(mockWaterline.catalogs.create).to.have.been.calledOnce;
+            });
+        });
+    });
+
+    describe('_handleAsyncRequest', function() {
+        beforeEach(function() {
+            WsmanInventoryJobClass.prototype._handleAsyncRequest.restore();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: what's purpose?

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+            wsmanInventoryJob.wsman = new wsmanTool(wsmanInventoryJob.dellConfigs.gateway, {
+                verifySSL: false,
+                recvTimeoutMs: 300000
+            });
+            wsmanTool.prototype.clientRequest.resolves({
+                body: {
+                    response: 'xxxxxx, Submitted: test, xxxxxx'
+                }
+            });
+            wsmanBaseJob.prototype._subscribeHttpResponseUuid.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: delete this line

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob._handleAsyncRequest().then(function() {
+                expect(wsmanTool.prototype.clientRequest).to.have.been.calledOnce;
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: expect logger.error not called?

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+            wsmanInventoryJob.wsman = new wsmanTool(wsmanInventoryJob.dellConfigs.gateway, {
+                verifySSL: false,
+                recvTimeoutMs: 300000
+            });
+            wsmanTool.prototype.clientRequest.resolves({
+                body: {
+                    response: 'xxxxxx, Submitted: test, xxxxxx'
+                }
+            });
+            wsmanBaseJob.prototype._subscribeHttpResponseUuid.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: delete this line

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+            return wsmanInventoryJob._handleAsyncRequest().then(function() {
+                expect(wsmanTool.prototype.clientRequest).to.have.been.calledOnce;
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: same problem

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+            wsmanInventoryJob.wsman = new wsmanTool(wsmanInventoryJob.dellConfigs.gateway, {
+                verifySSL: false,
+                recvTimeoutMs: 300000
+            });
+            wsmanTool.prototype.clientRequest.resolves({
+                body: {
+                    response: 'test'
+                }
+            });
+            wsmanBaseJob.prototype._subscribeHttpResponseUuid.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: delete this line

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+            wsmanInventoryJob.inventories = ['nics', 'manager'];
+            wsmanInventoryJob.dellConfigs = dellConfigs;
+            wsmanInventoryJob.nodeType = 'enclosure';
+            wsmanInventoryJob.nodeId = '59c4720870bf1d7c172930db';
+            wsmanInventoryJob.wsman = new wsmanTool(wsmanInventoryJob.dellConfigs.gateway, {
+                verifySSL: false,
+                recvTimeoutMs: 300000
+            });
+            wsmanTool.prototype.clientRequest.rejects();
+            wsmanBaseJob.prototype._subscribeHttpResponseUuid.resolves();
```
Comment (yaolingling) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: delete this line

Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 17 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 15 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 13 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 14 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 16 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 1 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 3 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 2 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 4 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 9 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 10 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 7 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done. Add new assertion to enhance validation.

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 8 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done. Add new assertion to enhance validation.

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 11 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Done

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+
+    afterEach(function() {
+        this.sandbox.restore();
+    });
+
+    describe('initJob', function() {
+        it('should handle the case of lacking dell configuration.', function() {
+            wsmanBaseJob.prototype.checkOBM.resolves(obm);
+            configuration.get.withArgs('dell').returns(undefined);
+            expect(wsmanInventoryJob._initJob()).to.be.rejectedWith('Dell Inventory web service is not defined in wsmanConfig.json.');
```
Comment (nortonluo) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: If there is no SMI service uri defined in smiConifg.json. What will happen?

File: spec/lib/jobs/dell-wsman-inventory-spec.js [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+
+var uuid = require('node-uuid');
+
+var obm = {
+    id: '59c4797e70bf1d7c172930dc',
+    node: '/api/2.0/nodes/59c4720870bf1d7c172930db',
+    service: 'dell-wsman-obm-service',
+    config: {
+      user: "admin",
+      host: "172.31.128.73"
```
Comment (nortonluo) [inline spec/lib/jobs/dell-wsman-inventory-spec.js]: No password?

File: spec/lib/jobs/dell-wsman-inventory-spec.js (hunk 19 above)
Comment (AlaricChan) [reply spec/lib/jobs/dell-wsman-inventory-spec.js]: Good suggestion. We could add some more tests to cover such cases later.

**Pull request merged**